### PR TITLE
528 dense foldr explicit implementation

### PR DESCRIPTION
**Pull request by djelovina**
some versions with monoids and/or operators might be still missing

### CODE REVIEW

File: include/alp/reference/blas1.hpp [hunk 1]
```diff
@@ -751,22 +767,42 @@ namespace alp {
...
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const Monoid & monoid = Monoid(),
-		const typename std::enable_if< ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value, void >::type * const = NULL ) {
+		const typename std::enable_if_t<
+			!alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value
+		> * const = NULL
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:776]: NULL -> nullptr

File: include/alp/reference/blas1.hpp [hunk 2]
```diff
@@ -751,22 +767,42 @@ namespace alp {
...
 		class Monoid
 	>
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const Monoid & monoid = Monoid(),
-		const typename std::enable_if< ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_monoid< Monoid >::value, void >::type * const = NULL ) {
+		const typename std::enable_if_t<
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:774]: I think the `typename` is not needed

File: include/alp/reference/blas1.hpp [hunk 3]
```diff
@@ -751,22 +767,42 @@ namespace alp {
...
+		> * const = NULL
+	) {
 		// static sanity checks
 		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D1, IOType >::value ), "alp::foldl",
-			"called with a vector x of a type that does not match the first domain "
+			"called with a vector y of a type that does not match the first domain "
 			"of the given operator" );
 		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename Monoid::D2, InputType >::value ), "alp::foldl",
 			"called on a vector y of a type that does not match the second domain "
 			"of the given operator" );
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:784]: I am not 100% sure about this, but should IOType match D2 and InputType match D1? This operation is doing IO = INPUT . IO

File: include/alp/reference/blas1.hpp [hunk 4]
```diff
@@ -779,23 +815,44 @@ namespace alp {
...
+		> * const = NULL
+	) {
 		// static sanity checks
-		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename OP::D1, IOType >::value ), "alp::foldl",
-			"called with a vector x of a type that does not match the first domain "
+		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename OP::D1, IOType >::value ), "alp::foldr",
+			"called with a vector y of a type that does not match the first domain "
 			"of the given operator" );
-		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename OP::D2, InputType >::value ), "alp::foldl",
+		NO_CAST_OP_ASSERT( ( ! ( descr & descriptors::no_casting ) || std::is_same< typename OP::D2, InputType >::value ), "alp::foldr",
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:830]: Same question about matching of data types with operator's D1 and D2 types.

File: include/alp/reference/blas1.hpp [hunk 5]
```diff
@@ -779,23 +815,44 @@ namespace alp {
...
 		class OP
 	>
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const OP & op = OP(),
-		const typename std::enable_if< ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_operator< OP >::value, void >::type * const = NULL ) {
+		const typename std::enable_if_t<
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:822]: typename not needed

File: include/alp/reference/blas1.hpp [hunk 6]
```diff
@@ -779,23 +815,44 @@ namespace alp {
...
+			!alp::is_object< InputType >::value && ! alp::is_object< IOType >::value && alp::is_operator< OP >::value
+		> * const = NULL
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:824]: NULL -> nullptr

File: include/alp/reference/blas1.hpp [hunk 7]
```diff
@@ -779,23 +815,44 @@ namespace alp {
 		typename IOType, typename IOStructure, typename IOView, typename IOImfR, typename IOImfC,
 		class OP
 	>
-	RC foldr( const Scalar< InputType, InputStructure, reference > & alpha,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Scalar< InputType, InputStructure, reference > &alpha,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const OP & op = OP(),
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:821]: no space between & and op

File: include/alp/reference/blas1.hpp [hunk 8]
```diff
@@ -875,10 +932,14 @@ namespace alp {
 		typename IOType, typename IOStructure, typename IOView, typename IOImfR, typename IOImfC,
 		class OP
 	>
-	RC foldr( const Vector< InputType, InputStructure, Density::Dense, InputView, InputImfR, InputImfC, reference > & x,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Vector< InputType, InputStructure, Density::Dense, InputView, InputImfR, InputImfC, reference > &x,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const OP & op = OP(),
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:938]: no space between & and op

File: include/alp/reference/blas1.hpp [hunk 9]
```diff
@@ -875,10 +932,14 @@ namespace alp {
...
-		const typename std::enable_if< alp::is_operator< OP >::value && ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value, void >::type * = NULL ) {
+		const typename std::enable_if_t<
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:939]: no typename

File: include/alp/reference/blas1.hpp [hunk 10]
```diff
@@ -875,10 +932,14 @@ namespace alp {
...
+			alp::is_operator< OP >::value && ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value
+		> * = NULL
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:941]: NULL -> nullptr

File: include/alp/reference/blas1.hpp [hunk 11]
```diff
@@ -890,53 +951,29 @@ namespace alp {
...
-		const size_t n = size( x );
-		if( n != size( y ) || n != size( m ) ) {
+		const size_t n = getLength( y );
+
+		if( getLength( x ) != n ) {
 			return MISMATCH;
 		}
 
-		throw std::runtime_error( "Needs an implementation." );
+		for ( size_t i = 0; i < n; ++i ) {
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:974]: no space between for and (

File: include/alp/reference/blas1.hpp [hunk 12]
```diff
@@ -1017,48 +1054,14 @@ namespace alp {
...
-	RC foldr( const Vector< InputType, InputStructure, Density::Dense, InputView, InputImfR, InputImfC, reference > & x,
-		const Vector< MaskType, MaskStructure, Density::Dense, MaskView, MaskImfR, MaskImfC, reference > & m,
-		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > & y,
+	RC foldr(
+		const Vector< InputType, InputStructure, Density::Dense, InputView, InputImfR, InputImfC, reference > &x,
+		Vector< IOType, IOStructure, Density::Dense, IOView, IOImfR, IOImfC, reference > &y,
 		const Monoid & monoid = Monoid(),
-		const typename std::enable_if< alp::is_monoid< Monoid >::value && ! alp::is_object< MaskType >::value && ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value, void >::type * =
-			NULL ) {
+		const typename std::enable_if_t<
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:1061]: no typename

File: include/alp/reference/blas1.hpp [hunk 13]
```diff
@@ -1017,48 +1054,14 @@ namespace alp {
...
+			alp::is_monoid< Monoid >::value && ! alp::is_object< InputType >::value && ! alp::is_object< IOType >::value
+		> * = NULL
```
Comment (vdimiczrc) [inline include/alp/reference/blas1.hpp:1063]: NULL -> nullptr

File: tests/unit/dense_fold.cpp [hunk 14]
```diff
@@ -43,54 +45,98 @@ void alp_program( const size_t &n, alp::RC &rc ) {
...
 
-		// test 1, check
+		// test 1 foldl, check
 #ifdef DEBUG
-		std::cout << "x = ";
+		std::cout << "x_l = ";
 #endif
-		for( size_t i = 0; i < alp::getLength( x ); ++i ) {
-			if( x[ i ] !=  testval1 * testval2 ) {
+		if ( *out_r != testval2 ) {
```
Comment (vdimiczrc) [inline tests/unit/dense_fold.cpp:69]: no space between if and (

File: tests/unit/dense_fold.cpp [hunk 15]
```diff
@@ -43,54 +45,98 @@ void alp_program( const size_t &n, alp::RC &rc ) {
...
 					  << " )\n";
 				rc = FAILED;
 				return;
 			}
 #ifdef DEBUG
 			if( i < 10 ) {
-				std::cout << x[ i ] << " ";
-			} else if ( i + 10 > alp::getLength( x ) ) {
-				std::cout << x[ i ] << " ";
+				std::cout << x_l[ i ] << " ";
```
Comment (vdimiczrc) [inline tests/unit/dense_fold.cpp:86]: no space between if and (. Also in two other occurrences below.

File: tests/unit/dense_fold.cpp [hunk 16]
```diff
@@ -43,54 +45,98 @@ void alp_program( const size_t &n, alp::RC &rc ) {
...
+					  << "vector [" <<  i << " ] ( " << x_r[ i ] << ", expected "
+					  << ( static_cast< T1 >( testval1 * testval2 ) )
+					  << " )\n";
+				rc = FAILED;
+				return;
+			}
+#ifdef DEBUG
+			if( i < 10 ) {
+				std::cout << x_r[ i ] << " ";
+			} else if ( i + 10 > alp::getLength( x_r ) ) {
```
Comment (vdimiczrc) [inline tests/unit/dense_fold.cpp:125]: no space between if and (

File: tests/unit/dense_fold.cpp [hunk 17]
```diff
@@ -131,25 +177,189 @@ void alp_program( const size_t &n, alp::RC &rc ) {
...
+					  << "vector [" <<  i << " ] ( " << x_l[ i ] << ", expected "
+					  << ( static_cast< T1 >( testval1 * testval2 ) )
+					  << " )\n";
+				rc = FAILED;
+				return;
+			}
+#ifdef DEBUG
+			if( i < 10 ) {
+				std::cout << x_l[ i ] << " ";
+			} else if ( i + 10 > alp::getLength( x_l ) ) {
```
Comment (vdimiczrc) [inline tests/unit/dense_fold.cpp:212]: spaces

File: tests/unit/dense_fold.cpp [hunk 18]
```diff
@@ -131,25 +177,189 @@ void alp_program( const size_t &n, alp::RC &rc ) {
...
```
Comment (vdimiczrc) [inline tests/unit/dense_fold.cpp:275]: spaces

File: include/alp/reference/blas1.hpp (hunk 3 above)
Comment (djelovina) [reply include/alp/reference/blas1.hpp:784]: yes since in op, D3 = D1.D2 and in foldr D2=D3 (in foldl D1=D3)

**Pull request merged**